zebra-network/src/meta_addr.rs

```diff
@@ -34,29 +112,48 @@ impl MetaAddr {
         let interval = crate::constants::TIMESTAMP_TRUNCATION_SECONDS;
         let ts = self.last_seen.timestamp();
         self.last_seen = Utc.timestamp(ts - ts.rem_euclid(interval), 0);
+        self.last_connection_state = Default::default();
         self
     }
 }
 
 impl Ord for MetaAddr {
-    /// `MetaAddr`s are sorted newest-first, and then in an arbitrary
-    /// but determinate total order.
+    /// `MetaAddr`s are sorted in approximate reconnection attempt order, but
+    /// with `Recent` peers sorted together.
+    ///
+    /// This order should not be used for reconnection attempts: use
+    /// [`AddressBook::reconnection_peers`] instead.
+    ///
+    /// See [`CandidateSet`] for more details.
```

Comment:
This ordering allows multiple `MetaAddr`s with the same address in the `BTreeMap`. But it seems like it's a source of subtle bugs and complexity.

We could:
- [ ] delete `Ord` and `PartialOrd` for `MetaAddr`
- [ ] remove the inner `BTreeMap`
- [ ] remove the consistency check function

As an alternative, we could document these issues, but the performance and code simplicity gain are probably worth it.

High priority performance follow up:
- [ ] limit the total address book size to maintain performance, discarding the oldest addresses once the limit is reached (2x the peer set target size should be more than sufficient for our needs - this is close to the current number of active mainnet peers, so the code actually gets exercised on a regular basis)

High priority security follow ups:
- [ ] handle future peer-provided last_seen times by subtracting the local/remote time delta from all the times in that peer's response
  - we use relative skew to avoid assuming anything about the correctness of the local or remote clocks (the security analysis and testing done absolute times can get quite complex)
  - [ ] open a follow-up to check for anywhere else in the protocol that uses peer-provided times or other untrusted/unverified data
- [ ] make future peer-provided last_seen times into a parse error, if possible
- [ ] filter out known addresses, then limit the number of new addresses used from each request to `target peer size/3` (with a minimum of 3, so we almost always get addresses from multiple nodes, because some nodes are dual stack)
  - this avoids having a single peer provide our entire set of peer connections (or our entire address book)
- [ ] in response to peer requests, only provide peers we've actually connected to (`AssumedAlive` or `AssumedDead`), to avoid gossiping unverified peer addresses, last_seen times, or services